### PR TITLE
parser+typechecker: Add support for types inside namespaces

### LINKED ISSUE #401: parser: Types cannot contain namespace qualifiers
Minimal repro:

```
namespace Foo {
    struct Bar { a: i32 }
}

function get_bar() -> Foo::Bar {
    return Foo::Bar(a: 10)
}
```

Produces: 

```
Error: expected complete block
---- file.jakt:5:26
 3 |  
 4 | function get_bar() -> Foo::Bar { 
                              ^- expected complete block
 5 |     return Foo::Bar(a: 10) 
 6 | } 
----
```

### PR DESCRIPTION
**Pull request by Baitinq**
Hi, this is my first jakt and rust related pull request so I might be doing something stupid:)

I believe this pull requests add support for properly parsing types inside a namespace in a function's return type.

I believe this fixes #401 . Now we get another error:

```
Error: Unknown type
---- file.jakt:5:23
 3 |
 4 | function get_bar() -> Foo::Bar {
                           ^- Unknown type
 5 |     return Foo::Bar(a: 10)
 6 | }
----
```

:^)

### CODE REVIEW

**Comment by MacDue** [comment]
Add a sample/test for this now you've added it

**Comment by jntrnr** [comment]
The way to do this would probably be to parse the type in the parser the same way we parse out a list of namespaces when we parse namespacing calls in other parts of the parser. Then later, to resolve that type to the ultimate TypeId in the typechecker

**Comment by Baitinq** [comment]
Update:

- Added tests
- Added support for types inside namespaces inside the typechecker

File: src/parser.rs [hunk 1]
```diff
@@ -4168,63 +4168,85 @@ pub fn parse_typename(tokens: &[Token], index: &mut usize) -> (ParsedType, Optio
...
+                    //Type in namespace
+                    *index += 1;
+                    let (inner_type, err) = parse_typename(tokens, index);
+                    if err.is_some() {
+                        error = error.or(err);
+                        return (unchecked_type, error)
+                    }
+                    if let ParsedType::Name(inner_type_name, _) = inner_type {
+                        unchecked_type = ParsedType::Name(
+                            typename + "::" + &inner_type_name,
```
Comment (jntrnr) [inline src/parser.rs:4239]: Can we give ParsedType a namespace specifier like we have for calls? That way when we resolve it we can walk the namespace path and find the type

Comment (Baitinq) [reply src/parser.rs:4239]: Sure, ill try and implement that today :)

**Comment by Baitinq** [comment]
@jntrnr I believe I have implemented your idea it after some struggle, please let me know if this works :^)

**Review by linusg — CHANGES_REQUESTED** [review]
`parser+formatter:`

**Comment by alimpfard** [comment]
- This doesn't allow types like `Foo::Bar<Baz>` (namespaced types with generics, for simplicity, only the last name can have generic params).
- Seems like some tests are failing, might be an actual issue, or might need a rebase.

**Comment by alimpfard** [comment]
Hey @Baitinq, I had to use your changes in my PR (#518), so I ended up moving your PR over there, could you take a look at that?
would it be okay to merge that change and close this PR?

**Comment by Baitinq** [comment]
Hi @alimpfard, yeah that looks great! I'll close this now:)

**Pull request closed without merging**